### PR TITLE
Migrate Upload Package To View Binding from Butterknife

### PR DESCRIPTION
**Pull request by shashankiitbhu**
**Description (required)**

Related to #4664 

What changes did you make and why?

**Tests performed (required)**

Tested {build variant, e.g. ProdDebug} on {name of device or emulator} with API level {API level}.

**Screenshots (for UI changes only)**

### CODE REVIEW

**Comment by neeldoshii** [comment]
Is this PR ready to review? @shashankiitbhu

File: app/src/main/java/fr/free/nrw/commons/upload/SimilarImageDialogFragment.java
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/SimilarImageDialogFragment.java:1]:
```
import android.widget.Button;
import com.facebook.drawee.view.SimpleDraweeView;
```
Consider removing this **Unused import statement**
Rest this package looks good.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java [hunk 1]
```diff
@@ -203,51 +205,53 @@ public void removeDescription(final UploadMediaDetail uploadMediaDetail, final i
...
         ImageView addButton;
 
-        @BindView(R.id.cl_parent)
         ConstraintLayout clParent;
 
-        @BindView(R.id.ll_write_better_caption)
         LinearLayout betterCaptionLinearLayout;
 
-        @BindView(R.id.ll_write_better_description)
         LinearLayout betterDescriptionLinearLayout;
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:226]: `TextInputeLayout, PasteSensitiveTextInputEditText, TextInputLayout, ImageView, ConstraintLayout, LinearLayout `

don't define variable for this its no longer needed in view binding it was needed in butterknife as we needed annotation. 
Remove this and its imports as well and optimize import.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java [hunk 2]
```diff
@@ -203,51 +205,53 @@ public void removeDescription(final UploadMediaDetail uploadMediaDetail, final i
...
 
+        private
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:228]: reformat & remove extra lines

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java [hunk 3]
```diff
@@ -203,51 +205,53 @@ public void removeDescription(final UploadMediaDetail uploadMediaDetail, final i
...
+            descInputLayout = binding.descriptionItemEditTextInputLayout;
+            captionItemEditText = binding.captionItemEditText;
+            captionInputLayout = binding.captionItemEditTextInputLayout;
+            removeButton = binding.btnRemove;
+            addButton = binding.btnAdd;
+            clParent = binding.clParent;
+            betterCaptionLinearLayout = binding.llWriteBetterCaption;
+            betterDescriptionLinearLayout = binding.llWriteBetterDescription;
+
+
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:254]: this all will be removed no need to store it in variable.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java [hunk 4]
```diff
@@ -495,7 +499,9 @@ private void onRecentLanguageClicked(final Dialog dialog, final AdapterView<?> a
             ((RecentLanguagesAdapter) adapterView
                 .getAdapter()).setSelectedLangCode(languageCode);
             Timber.d("Description language code is: %s", languageCode);
-            descriptionLanguages.setText(languageCode);
+            if (descriptionLanguages!=null) {
+                descriptionLanguages.setText(languageCode);
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:503]: binding. and access it

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java (hunk 3 above)
Comment (neeldoshii) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:254]: use binding. and access where ever the variable was used.

File: app/src/main/java/fr/free/nrw/commons/upload/categories/UploadCategoriesFragment.java [hunk 5]
```diff
@@ -340,9 +347,9 @@ public void onResume() {
         super.onResume();
 
         if (media != null) {
-            etSearch.setOnKeyListener((v, keyCode, event) -> {
+            binding.etSearch.setOnKeyListener((v, keyCode, event) -> {
                 if (keyCode == KeyEvent.KEYCODE_BACK) {
-                    etSearch.clearFocus();
+                    binding.etSearch.clearFocus();
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/categories/UploadCategoriesFragment.java:352]:
```
@Override
public void onDestroyView() {
    super.onDestroyView();
    binding = null; // Nullify the binding
}
``` 
missing

File: app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsFragment.java [hunk 6]
```diff
@@ -374,9 +373,9 @@ public void onResume() {
         super.onResume();
 
         if (media != null) {
-            depictsSearch.setOnKeyListener((v, keyCode, event) -> {
+            binding.depictsSearch.setOnKeyListener((v, keyCode, event) -> {
                 if (keyCode == KeyEvent.KEYCODE_BACK) {
-                    depictsSearch.clearFocus();
+                    binding.depictsSearch.clearFocus();
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/depicts/DepictsFragment.java:378]:
```
@Override
public void onDestroyView() {
    super.onDestroyView();
    binding = null; // Nullify the binding
}
```
missing

File: app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java [hunk 7]
```diff
@@ -26,13 +26,11 @@
...
 import androidx.recyclerview.widget.LinearLayoutManager;
 import androidx.recyclerview.widget.RecyclerView;
-import butterknife.BindView;
-import butterknife.ButterKnife;
-import butterknife.OnClick;
 import com.github.chrisbanes.photoview.PhotoView;
 import com.mapbox.mapboxsdk.camera.CameraPosition;
 import fr.free.nrw.commons.LocationPicker.LocationPicker;
 import fr.free.nrw.commons.R;
+import fr.free.nrw.commons.databinding.FragmentUploadMediaDetailFragmentBinding;
```
Comment (neeldoshii) [inline app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:33]: Remove unused import : 6 unused import

Comment (neeldoshii) [reply app/src/main/java/fr/free/nrw/commons/upload/mediaDetails/UploadMediaDetailFragment.java:33]: Same no one destroy fragment method needed.

File: app/src/test/kotlin/fr/free/nrw/commons/upload/categories/UploadCategoriesFragmentUnitTests.kt [hunk 8]
```diff
@@ -23,6 +23,7 @@ import fr.free.nrw.commons.OkHttpConnectionFactory
 import fr.free.nrw.commons.R
 import fr.free.nrw.commons.TestCommonsApplication
 import fr.free.nrw.commons.createTestClient
+import fr.free.nrw.commons.databinding.UploadCategoriesFragmentBinding
```
Comment (neeldoshii) [inline app/src/test/kotlin/fr/free/nrw/commons/upload/categories/UploadCategoriesFragmentUnitTests.kt:26]: Unused import optimized needed : 16 unused imports

**Review by neeldoshii — CHANGES_REQUESTED** [review]

**Comment by neeldoshii** [comment]
Left few minor code optimisation rest its good to go. I ran through the code locally. For unused optimisation fix this for this PR whichever files you touched. I will be adding klint #5198 in the project which will be ease in all this refactoring locally before commit. 

> Also just a feedback next time can you please create a package a PR.
>
> Few reasons:-
> 
> 1. This can lead to big workaround on merge conflict if someone else working on the same package and got things merged, also one single package its easy to get reviewed, fixing merge conflicts and faster merge.
> 
> 2. I don't have issue in reviewing multiple package at once but single package a pr would be best if someone (maintainer) if wishes to take a revert by some reason. Currently you have all the things in one commit. Reverting a commit will revert 19 packages to butterknifebinding whereas we just wanted to revert a package.
>Thanks.

Consider this as feedback.

Once the changes are done, its aprooved from my side.
Thank @shashankiitbhu 
**_Review by Student_**

File: app/src/test/kotlin/fr/free/nrw/commons/upload/categories/UploadCategoriesFragmentUnitTests.kt (hunk 8 above)
Comment (shashankiitbhu) [reply app/src/test/kotlin/fr/free/nrw/commons/upload/categories/UploadCategoriesFragmentUnitTests.kt:26]: this is unrelated to the issue it seems but I'll do it.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java (hunk 3 above)
Comment (shashankiitbhu) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:254]: Well, you haven't considered the case for :

@Override
    public void onViewAttachedToWindow(@NonNull final ViewHolder holder) {
        super.onViewAttachedToWindow(holder);
        holder.binding.captionItemEditText.setEnabled(false);
        holder.binding.captionItemEditText.setEnabled(true);

**Comment by shashankiitbhu** [comment]
> Also just a feedback next time can you please create a package a PR.

I added Per Package only in all of my PRs if you check carefully. Only one test is included in the PR which is related to that Package

File: app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java (hunk 1 above)
Comment (shashankiitbhu) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadMediaDetailAdapter.java:226]: Again
Screenshot 2024-03-09 at 1 44 05 PM

**Comment by neeldoshii** [comment]
> > Also just a feedback next time can you please create a package a PR.
> 
> I added Per Package only in all of my PRs if you check carefully. Only one test is included in the PR which is related to that Package

Apologies, my bad I took one file per PR.

**Pull request merged**